torch/_inductor/codegen/halide.py

```diff
@@ -0,0 +1,1121 @@
+                shape = None
+                dtype = "long"
+            else:
+                if arg.name in self.store_buffer_dimensions and "out" in arg.name:
+                    shape = [
+                        cexpr(self.rename_indexing(x))
+                        for x in self.store_buffer_dimensions[arg.name]
+                    ]
+                    assert shape
+                else:
+                    shape = [
+                        cexpr(
+                            self.rename_indexing(self.halide_buffer_numel(arg.buffer))
+                        )
+                    ] or ["1"]
+                dtype = f"{DTYPE_TO_CPP[arg.dtype]}*"
+            argtypes.append(
+                HalideInputSpec(
+                    dtype,
+                    arg.name,
+                    shape,
+                )
+            )
+        target = ["host", "strict_float"]
+        # TODO(jansel): for cuda want target="host-cuda-cuda_capability_86-user_context"
+        if not config.halide.asserts:
+            target.append("no_asserts")
+        if "64" in self.index_dtype:
+            # TODO(jansel): it is unclear if this does anything, since input sizes are still int32
+            target.append("large_buffers")
+
+        return HalideMeta(
+            argtypes,
+            target="-".join(target),
+            scheduler="Mullapudi2016",
+            scheduler_flags={
+                "parallelism": parallel_num_threads(),
+                "last_level_cache_size": HalideCodeCache.cpu_cache_size(),
+            },
+        )
+
+    def codegen_kernel(self, name=None):
+        """Called at the end to generate a final kernel string"""
+        if self.args.inplace_buffers:
+            raise Unsupported("inplace_buffers")
+        meta = self.halide_kernel_meta()  # ensure needed args are added early
+        code = IndentedBuffer()
+        code.splice(
+            """
+            import halide as hl
+
+            @hl.generator(name="kernel")
+            class Kernel:
+        """,
+            strip=True,
+        )
+        code.do_indent()
+        for _, arg in self.halide_argdefs():
+            if isinstance(arg, SizeArg):
+                code.writeline(f"{arg.name} = hl.InputScalar({self.index_dtype})")
+            else:
+                assert arg.buffer, arg
+                argcls = "hl.OutputBuffer" if "out" in arg.name else "hl.InputBuffer"
+                argtype = halide_type(arg.dtype)
+                ndim = len(self.store_buffer_dimensions.get(arg.name, (0,)))
+                code.writeline(f"{arg.name} = {argcls}({argtype}, {ndim})")
+        code.splice(
+            """
+            def generate(g):
+        """
+        )
+        code.do_indent()
+        for _, arg in self.halide_argdefs():
+            code.writeline(f"{arg.name} = g.{arg.name}")
+        for old, new in self.args.aliases():
+            code.writeline(f"{old} = {new}")
+
+        dom_size = {}
```

Comment:
Still need some time to review. But what the difference between the concept of 'Dom' here to a 'dimension' of a buffer?

Reply:
Halide has a concept of a RDom: https://halide-lang.org/docs/class_halide_1_1_r_dom.html

You use hl.RDom for reductions or updates and hl.Var for pure definitions.  RDom is more flexible, so I use that in some cases.